urllib3/util/rfc6555.py

```diff
@@ -0,0 +1,241 @@
+        family, expires = _HAPPY_EYEBALLS_CACHE[address]
+
+        # If the cache entry is expired, don't use it.
+        if start_time > expires:
+            del _HAPPY_EYEBALLS_CACHE[address]
+            family = 0
+        else:
+            cached = True
+
+    host, port = address
+    socks = []
+
+    # Make sure we close the selector after we're done.
+    with DefaultSelector() as selector:
+        # Perform a DNS lookup for the address for IPv4 or IPv6 families.
+        if not family:
+            family = socket.AF_UNSPEC
+
+        dns_results = socket.getaddrinfo(host, port, family, socket.SOCK_STREAM)
+        dns_results_len = len(dns_results)
+
+        for i in range(dns_results_len):
+            af, socktype, proto, canonname, sa = dns_results[i]
+
+            # We only care about IPv4 and IPv6 addresses.
+            if af not in _IP_FAMILIES:
+                continue
+
+            sock = None
+            try:
+                sock = socket.socket(af, socktype, proto)
+                if timeout_time is socket._GLOBAL_DEFAULT_TIMEOUT:
+                    sock_timeout = sock.gettimeout()
+                    if sock_timeout is None:
+                        timeout_time = None
+                    else:
+                        timeout_time = start_time + sock_timeout
+
+                if socket_options:
+                    for opt in socket_options:
+                        sock.setsockopt(*opt)
+
+                # If we're given a source address, bind to it.
+                if source_address:
+                    sock.bind(source_address)
+
+                # Set non-blocking for selecting.
+                sock.settimeout(0.0)
+
+                # Connect to the host.
+                errcode = sock.connect_ex(sa)
+                if errcode and errcode not in _ASYNC_ERROR_NUMBERS:
+                    err = socket.error(errcode)
+                    continue
+
+                errcode = sock.getsockopt(socket.SOL_SOCKET, socket.SO_ERROR)
+                if errcode and errcode not in _ASYNC_ERROR_NUMBERS:
+                    err = socket.error(errcode)
+                    continue
+
+                # Register this new socket with the selector.
+                socks.append(sock)
+
+                # Using EVENT_WRITE to detect a connection.
+                selector.register(sock, EVENT_WRITE)
+
+            except (socket.error, OSError) as e:
+                err = e
+                if sock is not None:
+                    sock.close()
+
+            # This is where the selecting logic occurs. Once all sockets
+            # have been created and are added to the selector, they should
+            # be continually selected until one of them works.  If there are
+            # still sockets to add to the selector, then select for only 200
+            # ms before adding the next socket to the selector.
+            reattempt_select = True
+            while reattempt_select:
```

Comment:
I have my suspicions about the correctness of this loop, but I can't easily tell if they're valid because the logic of this method is so complex.